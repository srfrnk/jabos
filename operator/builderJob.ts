import { ApiObjectProps, Chart } from 'cdk8s';
import { Container, EnvVar, KubeJob, KubeJobProps, Quantity, Volume } from './imports/k8s';
import jabosOperatorUrlEnv from './jabosOperatorUrlEnv';
import { GitRepositoryPropsEx, k8sName } from './misc';
import settings from './settings';

export default function (options: {
  object: ApiObjectProps,
  jobNamePrefix: string,
  imagePrefix: string,
  buildNumber: string,
  type: string,
  serviceAccountName: string,
  repo: GitRepositoryPropsEx,
  containers: Container[],
  volumes?: Volume[],
  metricName: string,
  metricLabels: { [key: string]: string; },
  labels: { [key: string]: string; },
  commitLabel: string
}): KubeJobProps {
  const latestCommit = options.repo.status.latestCommit;
  const jobName = k8sName(`${options.jobNamePrefix}-${options.object.metadata.name}`, latestCommit);
  const repoSsh = options.repo.spec.ssh;

  const initContainers: Container[] = (options.containers || []).map(container => {
    return {
      ...container,
      imagePullPolicy: settings.imagePullPolicy(),
      env: [
        ...(container.env || []),
        {
          name: "NAMESPACE",
          value: options.object.metadata.namespace
        },
        {
          name: "NAME",
          value: options.object.metadata.name
        },
        {
          name: "OBJECT_UID",
          value: options.object.metadata.uid
        },
      ],
      securityContext: {
        ...container.securityContext,
        readOnlyRootFilesystem: true,
        allowPrivilegeEscalation: false,
        runAsNonRoot: true,
        capabilities: {
          drop: ['ALL'],
        },
      },
      volumeMounts: [
        ...(container.volumeMounts || []),
        {
          name: "temp",
          mountPath: "/tmp",
        },
        {
          name: "build",
          mountPath: "/build",
        }
      ]
    };
  });

  const sshEnv: EnvVar[] = !repoSsh ? [] : [
    {
      name: "SSH_PASSPHRASE",
      valueFrom: {
        secretKeyRef: {
          name: repoSsh.secret,
          key: repoSsh.passphrase
        }
      }
    },
    {
      name: "SSH_KEY",
      valueFrom: {
        secretKeyRef: {
          name: repoSsh.secret,
          key: repoSsh.key,
        }
      }
    }
  ];

  return {
    metadata: {
      name: jobName,
      labels: options.labels
    },
    spec: {
      completions: 1,
      completionMode: "NonIndexed",
      backoffLimit: parseInt(settings.jobBackoffLimit()),
      activeDeadlineSeconds: parseInt(settings.jobActiveDeadlineSeconds()),
      ttlSecondsAfterFinished: parseInt(settings.jobTtlSecondsAfterFinished()),
      parallelism: 1,
      template: {
        metadata: {
          name: jobName,
          labels: {
            builder: jobName
          }
        },
        spec: {
          serviceAccountName: options.serviceAccountName,
          restartPolicy: "Never",
          securityContext: {
            runAsNonRoot: true,
          },
          initContainers: [
            {
              image: `${settings.imagePrefix()}pre-builder:${settings.buildNumber()}`,
              args: [],
              env: [
                {
                  name: "URL",
                  value: options.repo.spec.url
                },
                {
                  name: "BRANCH",
                  value: options.repo.spec.branch
                },
                {
                  name: "COMMIT",
                  value: latestCommit
                },
                {
                  name: "METRIC_NAME",
                  value: options.metricName
                },
                {
                  name: "METRIC_LABELS",
                  value: JSON.stringify(options.metricLabels)
                },
                {
                  name: "NAMESPACE",
                  value: options.object.metadata.namespace
                },
                {
                  name: "NAME",
                  value: options.object.metadata.name
                },
                {
                  name: "OBJECT_UID",
                  value: options.object.metadata.uid
                },
                ...jabosOperatorUrlEnv(),
                ...sshEnv
              ],
              imagePullPolicy: settings.imagePullPolicy(),
              securityContext: {
                readOnlyRootFilesystem: true,
                allowPrivilegeEscalation: false,
                runAsNonRoot: true,
                capabilities: {
                  drop: ['ALL'],
                },
              },
              name: "pre-builder",
              resources: {
                limits: {
                  cpu: Quantity.fromString("500m"),
                  memory: Quantity.fromString("500Mi")
                },
                requests: {
                  cpu: Quantity.fromString("100m"),
                  memory: Quantity.fromString("100Mi")
                }
              },
              volumeMounts: [
                {
                  name: "git-temp",
                  mountPath: "/gitTemp",
                },
                {
                  name: "temp",
                  mountPath: "/tmp",
                },
                {
                  name: "timer",
                  mountPath: "/timer",
                }
              ]
            },
            ...initContainers
          ],
          containers: [
            {
              image: `${options.imagePrefix}post-builder:${options.buildNumber}`,
              args: [],
              env: [
                {
                  name: "TYPE",
                  value: options.type
                },
                {
                  name: "NAME",
                  value: options.object.metadata.name
                },
                {
                  name: "COMMIT",
                  value: latestCommit
                },
                {
                  name: "COMMIT_LABEL",
                  value: options.commitLabel
                },
                {
                  name: "METRIC_NAME",
                  value: options.metricName
                },
                {
                  name: "METRIC_LABELS",
                  value: JSON.stringify(options.metricLabels)
                },
                {
                  name: "NAMESPACE",
                  value: options.object.metadata.namespace
                },
                {
                  name: "OBJECT_UID",
                  value: options.object.metadata.uid
                },
                ...jabosOperatorUrlEnv()
              ],
              volumeMounts: [
                {
                  name: "timer",
                  mountPath: "/timer",
                  readOnly: true
                },
                {
                  name: "temp",
                  mountPath: "/tmp"
                }
              ],
              imagePullPolicy: settings.imagePullPolicy(),
              securityContext: {
                readOnlyRootFilesystem: true,
                allowPrivilegeEscalation: false,
                runAsNonRoot: true,
                capabilities: {
                  drop: ['ALL'],
                },
              },
              name: "post-builder",
              resources: {
                limits: {
                  cpu: Quantity.fromString("500m"),
                  memory: Quantity.fromString("500Mi")
                },
                requests: {
                  cpu: Quantity.fromString("100m"),
                  memory: Quantity.fromString("100Mi")
                }
              },
            }
          ],
          volumes: [
            ...(options.volumes || []),
            {
              name: "git-temp",
              emptyDir: {}
            },
            {
              name: "temp",
              emptyDir: {}
            },
            {
              name: "build",
              emptyDir: {}
            },
            {
              name: "timer",
              emptyDir: {}
            }
          ]
        }
      }
    }
  };
}