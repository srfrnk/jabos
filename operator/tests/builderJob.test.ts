import { App, Chart } from 'cdk8s';
import builderJob from '../builderJob';
import { KubeJob } from '../imports/k8s';
import { clearMock, setMock } from './settingsMock';

beforeEach(() => {
  setMock();
});

afterEach(() => {
  clearMock();
});

test('awsSecretSources with options', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: { namespace: 'object_namespace', name: 'object_name', uid: 'object_uid' }
    },
    jobNamePrefix: 'jobNamePrefix_value',
    imagePrefix: 'imagePrefix_value-',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    commitLabel: 'builtCommit',
    serviceAccountName: 'serviceAccountName_value',
    repo: { spec: { url: 'url_value', branch: 'branch_value' }, status: { latestCommit: 'latestCommit_value' } },
    containers: [{ name: 'container_name' }],
    volumes: [{ name: 'volume_name' }],
    metricName: 'metricName_value',
    metricLabels: { 'metricLabelName': 'metricLabelValue' },
    labels: { 'label_name': 'label_value' }
  }
  const chart = new Chart(new App(), 'chart');
  expect(new KubeJob(chart, 'job', builderJob(options)).toJson()).toMatchSnapshot();
});

test('awsSecretSources with ssh', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: { namespace: 'object_namespace', name: 'object_name', uid: 'object_uid' }
    },
    jobNamePrefix: 'jobNamePrefix_value',
    imagePrefix: 'imagePrefix_value',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    commitLabel: 'builtCommit',
    serviceAccountName: 'serviceAccountName_value',
    repo: { spec: { url: 'url_value', branch: 'branch_value', ssh: { secret: 'secret_value', passphrase: 'passphrase_value', key: 'key_value' } }, status: { latestCommit: 'latestCommit_value' } },
    containers: [{ name: 'container_name' }],
    volumes: [{ name: 'volume_name' }],
    metricName: 'metricName_value',
    metricLabels: { 'metricLabelName': 'metricLabelValue' },
    labels: { 'label_name': 'label_value' }
  }
  const chart = new Chart(new App(), 'chart');
  expect(new KubeJob(chart, 'job', builderJob(options)).toJson()).toMatchSnapshot();
});

test('awsSecretSources with no containers', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: { namespace: 'object_namespace', name: 'object_name', uid: 'object_uid' }
    },
    jobNamePrefix: 'jobNamePrefix_value',
    imagePrefix: 'imagePrefix_value',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    commitLabel: 'builtCommit',
    serviceAccountName: 'serviceAccountName_value',
    repo: { spec: { url: 'url_value', branch: 'branch_value' }, status: { latestCommit: 'latestCommit_value' } },
    containers: null,
    volumes: [{ name: 'volume_name' }],
    metricName: 'metricName_value',
    metricLabels: { 'metricLabelName': 'metricLabelValue' },
    labels: { 'label_name': 'label_value' }
  }
  const chart = new Chart(new App(), 'chart');

  expect(new KubeJob(chart, 'job', builderJob(options)).toJson()).toMatchSnapshot();
});

test('awsSecretSources with no volumes', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: { namespace: 'object_namespace', name: 'object_name', uid: 'object_uid' }
    },
    jobNamePrefix: 'jobNamePrefix_value',
    imagePrefix: 'imagePrefix_value',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    commitLabel: 'builtCommit',
    serviceAccountName: 'serviceAccountName_value',
    repo: { spec: { url: 'url_value', branch: 'branch_value' }, status: { latestCommit: 'latestCommit_value' } },
    containers: [{ name: 'container_name' }],
    volumes: null,
    metricName: 'metricName_value',
    metricLabels: { 'metricLabelName': 'metricLabelValue' },
    labels: { 'label_name': 'label_value' }
  }
  const chart = new Chart(new App(), 'chart');
  expect(new KubeJob(chart, 'job', builderJob(options)).toJson()).toMatchSnapshot();
});

