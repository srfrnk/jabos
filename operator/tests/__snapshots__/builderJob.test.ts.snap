// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`awsSecretSources with no containers 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "label_name": "label_value",
    },
    "name": "jobNamePrefix_value-object_name-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "jobNamePrefix_value-object_name-latestCommit_value",
        },
        "name": "jobNamePrefix_value-object_name-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "builtCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "imagePrefix_valuepost-builder:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "url_value",
              },
              Object {
                "name": "BRANCH",
                "value": "branch_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "serviceAccountName_value",
        "volumes": Array [
          Object {
            "name": "volume_name",
          },
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;

exports[`awsSecretSources with no volumes 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "label_name": "label_value",
    },
    "name": "jobNamePrefix_value-object_name-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "jobNamePrefix_value-object_name-latestCommit_value",
        },
        "name": "jobNamePrefix_value-object_name-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "builtCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "imagePrefix_valuepost-builder:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "url_value",
              },
              Object {
                "name": "BRANCH",
                "value": "branch_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
          Object {
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
            ],
            "imagePullPolicy": "Always",
            "name": "container_name",
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "serviceAccountName_value",
        "volumes": Array [
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;

exports[`awsSecretSources with options 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "label_name": "label_value",
    },
    "name": "jobNamePrefix_value-object_name-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "jobNamePrefix_value-object_name-latestCommit_value",
        },
        "name": "jobNamePrefix_value-object_name-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "builtCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "imagePrefix_value-post-builder:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "url_value",
              },
              Object {
                "name": "BRANCH",
                "value": "branch_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
          Object {
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
            ],
            "imagePullPolicy": "Always",
            "name": "container_name",
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "serviceAccountName_value",
        "volumes": Array [
          Object {
            "name": "volume_name",
          },
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;

exports[`awsSecretSources with ssh 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "label_name": "label_value",
    },
    "name": "jobNamePrefix_value-object_name-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "jobNamePrefix_value-object_name-latestCommit_value",
        },
        "name": "jobNamePrefix_value-object_name-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "builtCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "imagePrefix_valuepost-builder:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "url_value",
              },
              Object {
                "name": "BRANCH",
                "value": "branch_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{\\"metricLabelName\\":\\"metricLabelValue\\"}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
              Object {
                "name": "SSH_PASSPHRASE",
                "valueFrom": Object {
                  "secretKeyRef": Object {
                    "key": "passphrase_value",
                    "name": "secret_value",
                  },
                },
              },
              Object {
                "name": "SSH_KEY",
                "valueFrom": Object {
                  "secretKeyRef": Object {
                    "key": "key_value",
                    "name": "secret_value",
                  },
                },
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
          Object {
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "object_namespace",
              },
              Object {
                "name": "NAME",
                "value": "object_name",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "object_uid",
              },
            ],
            "imagePullPolicy": "Always",
            "name": "container_name",
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "serviceAccountName_value",
        "volumes": Array [
          Object {
            "name": "volume_name",
          },
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;
